index.js

```diff
@@ -196,14 +196,28 @@
     return _funcPath(false, path, implementations);
   }
 
+  //  functionName :: Function -> String
+  var functionName = 'name' in function f() {} ?
+    function functionName(f) { return f.name; } :
+    /* istanbul ignore next */
+    function functionName(f) {
+      var match = /function (\w*)/.exec(f);
+      return match == null ? '' : match[1];
+    };
+
   //  $ :: (String, Array TypeClass, StrMap (Array Location)) -> TypeClass
   function $(_name, dependencies, requirements) {
     function getBoundMethod(_name) {
       var name = 'fantasy-land/' + _name;
       return requirements[_name] === Constructor ?
         function(typeRep) {
-          return funcPath([name], typeRep) ||
-                 implPath([/function (\w*)/.exec(typeRep)[1], name]);
+          var f = funcPath([name], typeRep);
+          if (f == null && typeof typeRep === 'function') {
+            var typeIdent = functionName(typeRep);
+            /* istanbul ignore else */
+            if (typeIdent !== '') f = implPath([typeIdent, name]);
```

Comment:
I'd like to test the else branch rather than ignore it. I haven't been able to come up with a test case, though, due to the extreme lengths that are taken to determine `Function#name`. For example:

```javascript
> ({constructor: function() {}}).constructor.name
'constructor'
```

If there isn't a scenario in which `typeIdent` will be `''` in an environment which provides `Function#name`, the guard isn't helpful. We could write this instead:

```javascript
var f = funcPath([name], typeRep);
return f == null && typeof typeRep === 'function' ?
  implPath([functionName(typeRep), name]) :
  f;
```